### PR TITLE
Add more span-based Vector<T> ctors and CopyTo methods (ref asm + unit tests)

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/corefx/issues/30968.

This PR contains the reference APIs and unit tests for https://github.com/dotnet/coreclr/pull/23333. CI will fail until that PR makes its way through coreclr and over here to corefx.

### CODE REVIEW

File: src/System.Numerics.Vectors/ref/System.Numerics.Vectors.cs [hunk 1]
```diff
@@ -471,6 +471,8 @@ public partial struct Vector<T> : System.IEquatable<System.Numerics.Vector<T>>,
     {
         private int _dummyPrimitive;
 #if HAS_SPAN
```
Comment (stephentoub) [inline src/System.Numerics.Vectors/ref/System.Numerics.Vectors.cs]: This predates your PR, but I thought we generally separated .NET Core additions like this into a separate .netcoreapp.cs file when we need to build the ref for both old and new surface areas?

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs [hunk 2]
```diff
@@ -212,15 +284,121 @@ public partial class GenericVectorTests
...
+        public void CopyToSpanUInt64() { TestCopyToSpan<ulong>(); }
+        [Fact]
+        public void CopyToSpanInt64() { TestCopyToSpan<long>(); }
+        [Fact]
+        public void CopyToSpanSingle() { TestCopyToSpan<float>(); }
+        [Fact]
+        public void CopyToSpanDouble() { TestCopyToSpan<double>(); }
+        private void TestCopyToSpan<T>() where T : struct
+        {
+            var initialValues = GenerateRandomValuesForVector<T>();
```
Comment (stephentoub) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: Nit: var => whatever type it actually is

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs [hunk 3]
```diff
@@ -212,15 +284,121 @@ public partial class GenericVectorTests
...
+        public void TryCopyToSpanUInt64() { TestTryCopyToSpan<ulong>(); }
+        [Fact]
+        public void TryCopyToSpanInt64() { TestTryCopyToSpan<long>(); }
+        [Fact]
+        public void TryCopyToSpanSingle() { TestTryCopyToSpan<float>(); }
+        [Fact]
+        public void TryCopyToSpanDouble() { TestTryCopyToSpan<double>(); }
+        private void TestTryCopyToSpan<T>() where T : struct
+        {
+            var initialValues = GenerateRandomValuesForVector<T>();
```
Comment (stephentoub) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: Ditto

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs [hunk 4]
```diff
@@ -212,15 +284,121 @@ public partial class GenericVectorTests
...
+            var vector = new Vector<T>(initialValues);
+            Span<T> destination = new T[Vector<T>.Count];
+
+            Assert.False(vector.TryCopyTo(destination.Slice(1)));
```
Comment (stephentoub) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: Do we want to validate anything about the contents of destination after the copy fails?  I'm ok if the answer is "it may or may not have written data", just checking.

**Review by stephentoub — APPROVED** [review]

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: I was just talking to @terrajobst about this. @terrajobst, see the use of `var` here, which increases the code review difficulty. If you had to guess, from local context, what type do you think is `initialValues`?

<details>
  <summary>Answer:</summary>

https://github.com/dotnet/corefx/blob/82c6f6320aeb14186f0438e77bfdc383eb631d05/src/System.Numerics.Vectors/tests/Performance/Constructor/GenericVectorConstructorTests.tt#L88

</details>

Comment (GrabYourPitchforks) [reply src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: I wanted to change as little as possible in the existing code, even if it meant propagating undesirable coding conventions. If this is causing substantial agitation for reviewers I can change the .tt file all the way up and down as part of this PR. :)

File: src/System.Numerics.Vectors/ref/System.Numerics.Vectors.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/System.Numerics.Vectors/ref/System.Numerics.Vectors.cs]: I think you're right. This is the only place across corefx where I see `HAS_SPAN` in any reference assembly project.

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: > If this is causing substantial agitation for reviewers I can change the .tt file all the way up and down as part of this PR. 

Yes please :)  But since this code in general predated your PR (right?), it can be a separate PR to just fix that up.

**Comment by tannergooding** [comment]
CC. @GrabYourPitchforks

Is this going in?

**Comment by GrabYourPitchforks** [comment]
Yes, but not for P5.

**Comment by GrabYourPitchforks** [comment]
Finally getting around to some older PRs now that all the fires are put out. :) The easiest way to review the changes for the most current iteration would be to look at the single commit that's responding to PR feedback: https://github.com/dotnet/corefx/pull/36136/commits/0444f0077c79ce4794f7d6f8a63ec0e46266e349. I think I addressed everybody's feedback there.

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs [hunk 5]
```diff
@@ -43,14 +44,85 @@ public partial class GenericVectorTests
...
+            // Span<T> ctor
+            vector = new Vector<T>(new Span<T>(values));
             ValidateVector(vector,
                 (index, val) =>
                 {
                     Assert.Equal(values[index], val);
                 });
+
+            // ReadOnlySpan<T> ctor
+            vector = new Vector<T>(new ReadOnlySpan<T>(values));
```
Comment (tannergooding) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: nit: Might be nice to have these as separate unit tests, since are testing different inputs. Otherwise, one failure prevents the other checks from happening.

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs [hunk 6]
```diff
@@ -43,14 +44,85 @@ public partial class GenericVectorTests
...
+            ValidateVector(vector,
+                (index, val) =>
+                {
+                    Assert.Equal(values[index], val);
+                });
+
+            // ReadOnlySpan<byte> ctor
+            vector = new Vector<T>(MemoryMarshal.AsBytes(new ReadOnlySpan<T>(values)));
```
Comment (tannergooding) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: Is this validation path particularly useful? It seems like if taking a `Span` works, it should work and things like `MemoryMarshal.AsBytes` returning a functioning Span should have been tested as part of `MemoryMarshal`?

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs [hunk 7]
```diff
@@ -43,14 +44,85 @@ public partial class GenericVectorTests
...
+            ValidateVector(vector,
+                (index, val) =>
+                {
+                    Assert.Equal(values[index], val);
+                });
+        }
+
+        [Fact]
+        public void ReadOnlySpanBasedConstructorWithLessElements_Byte() => Assert.Throws<IndexOutOfRangeException>(() => TestReadOnlySpanBasedConstructorWithLessElements<byte>());
```
Comment (tannergooding) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: Seems like having a customized `Theory` attribute that took a set of `T` as part of the `InlineData` would be useful...

Comment (tannergooding) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: Might be also nice to explicitly test an empty `Span`.

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs [hunk 8]
```diff
@@ -212,15 +284,129 @@ public partial class GenericVectorTests
         [Fact]
         public void ConstructorWithUnsupportedTypes_DateTime() => TestConstructorWithUnsupportedTypes<DateTime>();
         [Fact]
-        public void ConstructorWithUnsupportedTypes_Char() => TestConstructorWithUnsupportedTypes<char>();
+        public void ConstructorWithUnsupportedTypes_Char() => TestConstructorWithUnsupportedTypes<Char>();
```
Comment (tannergooding) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: Why the change to uppercase `char`? (also, `bool` is another good one to test)

Comment (GrabYourPitchforks) [reply src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: The TT somehow started emitting the actual type names instead of the friendly names so I had to make some tweaks to the TT itself. I thought I had fixed all occurrences but apparently I missed one. :)

File: src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs (hunk 7 above)
Comment (tannergooding) [reply src/System.Numerics.Vectors/tests/GenericVectorTests.netcoreapp.cs]: (to clarify, not saying this needs to be done here/now; just that it might be useful in the future)

**Comment by karelz** [comment]
@GrabYourPitchforks no update for 2 weeks. What is the plan for this PR?

**Comment by GrabYourPitchforks** [comment]
Tanner already updated the ref asms in https://github.com/dotnet/corefx/pull/37473 while I was OOF. Moving this PR to Future since it's just unit tests and needs a little bit of cleanup.

**Comment by karelz** [comment]
@GrabYourPitchforks waht's the plan here? Another 5 weeks passed. Now that master is opened for 5.0, we should be able to get it in, right?

**Comment by danmosemsft** [comment]
I would clsoe this for age but the feature is already in, so we need the tests 😃
